### PR TITLE
Apply PHP 7.4 syntax and typed property

### PR DESCRIPTION
**Pull request by samsonasik**
Signed-off-by: Abdul Malik Ikhsan <samsonasik@gmail.com>

|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

Since composer.json require php 7.4, php 7.4 syntax can be applied with typed properties. 

- For `final class`, typed properties for protected is allowed if no inherit of parent or used by its final class parent.
- Otherwise, update only private properties.

### CODE REVIEW

File: src/Http/Chain.php [hunk 1]
```diff
@@ -164,10 +163,8 @@ public function assemble(array $params = [], array $options = [])
 
         $this->assembledParams = [];
 
-        $routes = ArrayUtils::iteratorToArray($this->routes);
-
-        end($routes);
-        $lastRouteKey = key($routes);
+        $routes       = ArrayUtils::iteratorToArray($this->routes);
+        $lastRouteKey = array_key_last($routes);
```
Comment (Ocramius) [inline src/Http/Chain.php:167]: Oh nice!

**Review by Ocramius — APPROVED** [review]

**Comment by Ocramius** [comment]
Thanks @samsonasik!

File: src/Http/Chain.php (hunk 1 above)
Comment (Ocramius) [reply src/Http/Chain.php:167]: If this was done by rector: impressive!

Comment (samsonasik) [reply src/Http/Chain.php:167]: yes 👍 , there is rule for that `ArrayKeyFirstLastRector` https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#arraykeyfirstlastrector

**Pull request merged**